scanner/decisions/0002-local-scanning-with-scanner-v4.md

```diff
@@ -0,0 +1,64 @@
+| GetImageVulnerabilities | Matcher/GetVulnerabilities | Retrieve the matching vulnerabilities for the components provided.    |
+
+Although `GetImageVulnerabilities` and `Matcher/GetVulnerabilities` are similar, they have a crucial difference. `Matcher/GetVulnerabilities` was specifically designed to retrieve the index report to be scanned from a separate service or storage, similar to how [Clair is implemented](https://github.com/quay/clair/blob/main/httptransport/matcher_v1.go#L116). On the other hand, `GetImageVulnerabilities` accepts a list of components in its payload. To fully utilize the clear separation between Indexer and Matcher in Scanner V4's design, changes need to be made to Scanner V4, Central, and Sensor. This includes Central and Sensor's ability to handle Index Reports and also Scanner V4's capability to access reports that are generated in the secured cluster.
+
+## Decision
+
+Scanner V4's Matcher will accept a new parameter in `GetVulnerabilitiesRequest` to support local scanning. This parameter will allow clients to provide the `v4.IndexReport` for scanning. The Matcher will then verify if the report components are available in the request. If not, it will retrieve them from the `Indexer`.
+
+Example:
+
+```proto
+message GetVulnerabilitiesRequest {
+        string hash_id = 1;
+        Contents contents = 2;
+}
+```
+
+Notice that we don't necessarily need to embed the whole Index Report since the Matcher only needs a subset of the proto fields:
+
+```proto
+message Contents {
+    repeated Package              packages      = 1;
+    repeated Distribution         distributions = 2;
+    repeated Repository           repositories  = 3;
+    map<string, Environment.List> environments  = 4;
+}
+```
+
+But that will be left to the implementation to decide.
+
+Sensor and Central will behave the same as today regarding handling scan requests. Scanner V4 clients will have to make minimal changes to the interfaces of current scanner clients[^1] to support both V2 and V4. In central, a new field for the `IndexreportComponents` will be added [to the `EnrichLocalImageInternalRequest`](https://github.com/stackrox/stackrox/blob/a21793de1842586499e4afb3de68b780753db7f0/proto/api/v1/image_service.proto#L62).
+
+Finally, Index Reports generated in the secured cluster will not be stored in the central cluster. They will be stored in the secured cluster, in Scanner DB.
+
+## Consequences
+
+1. This method eliminates the need to store Index Reports in the central cluster for images obtained from the secured cluster. Instead, it distributes the workload of adding, storing, and retrieving Index Reports in the central Scanner's database across a large number of secured clusters connected to the Central.
+2. Re-scanning continues to rely on Deployment Resyncs.  That means images are only rescanned after a resync deployment message is sent to Sensor, and the image cache has expired.
```

Comment:
I believe this line is referring to the `ReprocessDeployments` MsgToSensor? If so would remove "resync" as that may be confused with informer resyncs.  Believe the statement will remain accurate if relying on informer resyncs or not.  The `ReprocessDeployments` message will either clear the `deduper` cache triggering reprocessing on next resync or submit all the deployments for reprocessing if resync disabled.

I believe the strategy is to remove reliance on resyncs within Sensor FYI.

Reply:
That makes sense. I will update the text to make it clear the role of the `ReprocessDeployments` vs. Informer Resyncs.